src/simulators/circuit_executor.hpp

```diff
@@ -0,0 +1,1174 @@
+/**
+ * This code is part of Qiskit.
+ *
+ * (C) Copyright IBM 2018, 2019. 2023.
+ *
+ * This code is licensed under the Apache License, Version 2.0. You may
+ * obtain a copy of this license in the LICENSE.txt file in the root directory
+ * of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Any modifications or derivative works of this code must retain this
+ * copyright notice, and modified files need to carry a notice indicating
+ * that they have been altered from the originals.
+ */
+
+#ifndef _circuit_executor_hpp_
+#define _circuit_executor_hpp_
+
+#include "framework/config.hpp"
+#include "framework/creg.hpp"
+#include "framework/json.hpp"
+#include "framework/opset.hpp"
+#include "framework/results/experiment_result.hpp"
+#include "framework/results/result.hpp"
+#include "framework/rng.hpp"
+#include "framework/types.hpp"
+#include "noise/noise_model.hpp"
+
+#include "transpile/cacheblocking.hpp"
+#include "transpile/fusion.hpp"
+
+#include "simulators/simulators.hpp"
+#include "simulators/state.hpp"
+
+namespace AER {
+
+namespace CircuitExecutor {
+
+using OpItr = std::vector<Operations::Op>::const_iterator;
+
+// Timer type
+using myclock_t = std::chrono::high_resolution_clock;
+
+//-------------------------------------------------------------------------
+// Executor base class
+//-------------------------------------------------------------------------
+template <class state_t>
+class Executor {
+protected:
+  // Simulation method
+  Method method_;
+
+  // Simulation device
+  Device sim_device_ = Device::CPU;
+
+  // Simulation precision
+  Precision sim_precision_ = Precision::Double;
+
+  // Save counts as memory list
+  bool save_creg_memory_ = false;
+
+  // The maximum number of threads to use for various levels of parallelization
+  int max_parallel_threads_;
+
+  // Parameters for parallelization management in configuration
+  int max_parallel_shots_;
+  size_t max_memory_mb_;
+  size_t max_gpu_memory_mb_;
+  int num_gpus_; // max number of GPU per process
+
+  // use explicit parallelization
+  bool explicit_parallelization_;
+
+  // Parameters for parallelization management for experiments
+  int parallel_experiments_;
+  int parallel_shots_;
+  int parallel_state_update_;
+
+  // results are stored independently in each process if true
+  bool accept_distributed_results_ = true;
+
+  uint_t myrank_;               // process ID
+  uint_t nprocs_;               // number of processes
+  uint_t distributed_rank_;     // process ID in communicator group
+  uint_t distributed_procs_;    // number of processes in communicator group
+  uint_t distributed_group_;    // group id of distribution
+  int_t distributed_proc_bits_; // distributed_procs_=2^distributed_proc_bits_
+                                // (if nprocs != power of 2, set -1)
```

Comment:
```
  uint_t myrank_;               // process ID
  uint_t nprocs_;               // number of processes
  uint_t distributed_rank_;     // process ID in communicator group
  uint_t distributed_procs_;    // number of processes in communicator group
  uint_t distributed_group_;    // group id of distribution
  int_t distributed_proc_bits_; // distributed_procs_=2^distributed_proc_bits_
                                // (if nprocs != power of 2, set -1)
```
Theses values are only for MPI, I think. Can we put them in `#ifdef AER_MPI`. Maybe we set some codes in `aer_controller.hpp` in it.

Reply:
There are many depending codes to these params, so I would like to keep them available for non MPI environment. But I added `#ifdef AER_MPI `to the codes we can separate.